pkg/certificate/mrc_reconciler.go

```diff
@@ -1,214 +1,182 @@
-// determine if we can propagate the rotation forward.
-type UseCase string
-
-const (
-	// UseCaseValidatingWebhook maps to the certificate use case for the validating webhook.
-	UseCaseValidatingWebhook UseCase = "validatingWebhook"
-
-	// UseCaseMutatingWebhook maps to the certificate use case for the Mutating Webhook
-	UseCaseMutatingWebhook UseCase = "mutatingWebhook"
-
-	// UseCaseXDSControlPlane maps to the certificate use case for the XDS Control Plane
-	UseCaseXDSControlPlane UseCase = "xdsControlPlane"
+func (m *Manager) handleMRCEvent(event MRCEvent) error {
+	log.Debug().Msgf("handling MRC event for MRC %s", event.MRCName)
+	mrcList, err := m.mrcClient.ListMeshRootCertificates()
+	if err != nil {
+		return err
+	}
 
-	// UseCaseSidecar maps to the certificate use case for the  sidecar service certs
-	UseCaseSidecar UseCase = "sidecar"
+	if len(mrcList) == 0 {
+		msg := fmt.Sprintf("when handling MRC event for MRC %s, found no MRCs in OSM control plane namespace", event.MRCName)
+		log.Error().Msg(msg)
+		return fmt.Errorf(msg)
+	}
 
-	// UseCaseBootstrap maps to the certificate use case for the  Bootstrap secrets
-	UseCaseBootstrap UseCase = "bootstrap"
+	filteredMRCList := filterOutInactiveMRCs(mrcList)
+	if len(filteredMRCList) > 2 {
+		log.Error().Err(ErrNumMRCExceedsMaxSupported).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrNumMRCExceedsMaxSupported)).
+			Msgf("expected 2 or less MRCs in the OSM control plane namespace, found %d", len(mrcList))
+		return ErrNumMRCExceedsMaxSupported
+	}
 
-	// UseCaseGateway maps to the certificate use case for the  Gateway ingress cert.
-	UseCaseGateway UseCase = "gateway"
-)
+	var mrc1, mrc2 *v1alpha2.MeshRootCertificate
+	mrc1 = mrcList[0]
+	if len(filteredMRCList) == 1 {
+		return m.handleSingleMRC(mrc1)
+	}
 
-var (
-	allUseCases = []UseCase{
-		UseCaseValidatingWebhook,
-		UseCaseMutatingWebhook,
-		UseCaseXDSControlPlane,
-		UseCaseSidecar,
-		UseCaseBootstrap,
-		UseCaseGateway,
+	mrc2 = mrcList[1]
+	log.Debug().Msg("validating MRC intent combination")
+	if err = ValidateMRCIntents(mrc1, mrc2); err != nil {
+		return err
 	}
-)
 
-func (uc UseCase) String() string {
-	return string(uc)
-}
-func (m *Manager) handleMRCEvent(event MRCEvent) error {
-	mrc := event.MRC
-	// TODO(5046): remove this first call to setIssuers.
-	err := m.setIssuers(event.MRC)
-	if err == nil {
-		return nil
-	}
-	log.Err(err).Msg("error setting issuers on mrc")
-	// TODO(5046): improve logging in this method.
-	if shouldUpdateMRCComponentStatus(mrc) {
-		if err := m.updateMRCComponentStatus(mrc); err != nil {
-			return err
-		}
-	} else if shouldUpdateState(mrc) {
-		if err := m.updateMRCState(mrc); err != nil {
-			return err
-		}
-	} else if m.shouldSetIssuers(mrc) {
-		log.Info().Msgf("setting new certificate issuers on the Certificate Manager for MRC %s in stage %s", mrc.GetName(), mrc.Status.State)
-		return m.setIssuers(mrc)
+	if err = m.setIssuers(mrc1, mrc2); err != nil {
```

Comment:
can we add a check to see if we should be setting the issuers? ie: if the mrc hasn't changed since the previously set mrc's, then we shouldn't set them.

Calling `getCertIssuer` calls `GetCertIssuerForMRC`, which makes an API call, so could be expensive if we do it repeatedly. And indeed, might cost the customer money depending on their PKI infrastructure billing plan.

Reply:
Added `shouldSetIssuer` that we be called before `getCertIssuer` and `GetCertIssuerForMRC`.